shared/utils/logic.ts

```diff
@@ -0,0 +1,336 @@
+  return formLogic.logicType === LogicType.ShowFields
+}
+
+/**
+ * Checks if the field ids in logic's conditions all exist in the fieldIds.
+ * @param conditions the list of conditions to check
+ * @param formFieldIds the set of form field ids to check
+ * @returns true if every condition's related form field id exists in the set of formFieldIds, false otherwise.
+ */
+const allConditionsExist = (
+  conditions: FormCondition[],
+  formFieldIds: FieldIdSet,
+): boolean => {
+  return conditions.every((condition) =>
+    formFieldIds.has(String(condition.field)),
+  )
+}
+
+/**
+ * Parse logic to get a list of conditions where, if any condition in this list
+ * is fulfilled, form submission is prevented.
+ * @param form the form document to check
+ * @returns array of conditions that prevent submission, can be empty
+ */
+const getPreventSubmitConditions = (
+  form: PickLogicSubset,
+): PreventSubmitLogic[] => {
+  const formFieldIds = new Set(
+    form.form_fields.map((field) => String(field._id)),
+  )
+
+  const preventFormLogics =
+    form.form_logics?.filter(
+      (formLogic): formLogic is PreventSubmitLogic =>
+        isPreventSubmitLogic(formLogic) &&
+        allConditionsExist(formLogic.conditions, formFieldIds),
+    ) ?? []
+
+  return preventFormLogics
+}
+
+/**
+ * Parse logic into a map of fields that are shown/hidden depending on the
+ * values of other fields.
+ * Discards invalid logic, where the id in show or conditions do not exist in
+ * the form_field.
+ *
+ * @example
+ * Show Email (_id: 1001) and Number (_id: 1002) if Dropdown (_id: 1003) is "Option 1" and Yes_No (_id: 1004) is "Yes"
+  Then,
+  form_logics: [
+    {
+      show: ["1001","1002"],
+      conditions: [
+        {field: "1003", ifValueType: "single-select", state: "is equals to", value: "Option 1"},
+        {field: "1004", ifValueType: "single-select", state: "is equals to", value: "Yes"}
+       ]
+    }
+  ]
+
+  logicUnitsGroupedByField:
+  {
+    "1001": [ [{field: "1003", ifValueType: "single-select", state: "is equals to", value: "Option 1"},
+        {field: "1004", ifValueType: "single-select", state: "is equals to", value: "Yes"}] ],
+    "1002": [ [{field: "1003", ifValueType: "single-select", state: "is equals to", value: "Option 1"},
+        {field: "1004", ifValueType: "single-select", state: "is equals to", value: "Yes"}] ]
+  }
+ * @caption If "1001" is deleted, "1002" will still be rendered since we just won't add "1001" into logicUnitsGroupedByField
+ *
+ * @param form the form object to group its logic by field for
+ * @returns an object containing fields to be displayed and their corresponding conditions, keyed by id of the displayable field
+ */
+export const groupLogicUnitsByField = (form: PickLogicSubset): GroupedLogic => {
+  const formId = form._id
+  const formLogics = form.form_logics.filter(isShowFieldsLogic) ?? []
+
+  const formFieldIds = new Set(
+    form.form_fields?.map((field) => String(field._id)),
```

Comment:
I think can remove the optional chain since form_fields now (correctly) is defined